### DISCUSSION TITLE
tun2socks in rust?

### DISCUSSION THREAD

**Discussion by Mygod** [shadowsocks/shadowsocks-android#2721] [Ideas]
So... are we going to replace tun2socks with rust too? What is the status on this? https://github.com/shadowsocks/tun2socks

CC: @madeye @zonyitoo @LuoZijun 😛

**Comment by zonyitoo** [comment]
I was waiting for @LuoZijun to make it stable and then I could integrate some of it into shadowsocks-rust.

I recently found that the most convenient way to use shadowsocks was iptables, which is why I didn't try to push it forward.

**Reply by Mygod** [reply]
Well too bad Android couldn't use iptables... (see the other thread)

**Comment by LuoZijun** [comment]
@Mygod `tun2socks` 主要是作为一个样板代码来演示使用 `smoltcp` 取代 `lwip` 是可行的。

至于它什么时候能够作为组件进入 `ss-rust` 里面去，来取代繁琐无聊的 `iptables` 以及 `CN-IP` 相关的命令，我想或许不是很重要？这主要看 @zonyitoo 的意愿。

至于我个人，我希望它会在下一代的 `ss` 协议里面，暂且就叫它 `v2` 好了。这个 `v2` 的实现，肯定会包含 `tun2socks`、`GeoDNS`、`TLSv1.3-Like-Protocol`。

下面的这些项目都只是实现 `v2` 时，做的一些技术验证：

* https://github.com/shadowsocks/tls
* https://github.com/shadowsocks/iana-ip-db
* https://github.com/shadowsocks/tun2socks
* https://github.com/shadowsocks/dns
* https://github.com/shadowsocks/crypto2

至于 https://github.com/shadowsocks/shadowsocks-crypto 这是另外的产物，它被使用在了当前的 `ss-rust` v1 版本里面。

目前，还有 2 个部分没有完成：

* DoT
* TLS-v1.3

这两个部分完成之后，我会提出 `v2` 的协议供大家讨论，同时会有一个 `ss-v2` 的 demo。

NOTE: 至于一些库的实现选择重新造轮子，是因为我对目前 Rust 生态里面广泛使用的库的质量并不满意，如 `trust-dns`、`rustls`。
